surveyor/observation.go

```diff
@@ -317,407 +296,329 @@ func (o *ServiceObsListener) Stop() {
-		s.logger.Debugf("starting listener goroutine for %s", dir)
-		for {
-			select {
-			case event, ok := <-watcher.Events:
-				if !ok {
-					return
-				}
-				if err := s.handleWatcherEvent(event, depth); err != nil {
-					s.logger.Warn(err)
-				}
-			case <-s.stop:
-				return
-			}
-		}
-	}()
-	return nil
+	om.running = true
 }
 
-func (s *Surveyor) handleWatcherEvent(event fsnotify.Event, depth int) error {
-	path := event.Name
-	s.Lock()
-	defer s.Unlock()
+// IsRunning returns true if the observation manager is running or false if it is stopped
+func (om *ServiceObsManager) IsRunning() bool {
+	om.Lock()
+	defer om.Unlock()
+	return om.running
+}
 
-	switch {
-	case event.Has(fsnotify.Create):
-		return s.handleCreateEvent(path, depth)
-	case event.Has(fsnotify.Write) && !event.Has(fsnotify.Remove):
-		return s.handleWriteEvent(path)
-	case event.Has(fsnotify.Remove):
-		return s.handleRemoveEvent(path)
+func (om *ServiceObsManager) stop() {
+	om.Lock()
+	defer om.Unlock()
+	if !om.running {
+		return
 	}
-	return nil
-}
 
-func (s *Surveyor) handleCreateEvent(path string, depth int) error {
-	stat, err := os.Stat(path)
-	if err != nil {
-		return fmt.Errorf("could not read observation file %s: %s", path, err)
+	for _, obs := range om.listenerMap {
+		obs.Stop()
 	}
-	// if a new directory was created, first start all observations already in it
-	// and then start watching for changes in this directory (fsnotify.Watcher is not recursive)
-	if stat.IsDir() && !strings.HasPrefix(stat.Name(), "..") {
-		depth--
-		err = filepath.WalkDir(path, s.startObservationsInDir())
-		if err != nil {
-			return fmt.Errorf("could not start observation from %s: %s", path, err)
-		}
-		if err := s.watchObservations(path, depth); err != nil {
-			return fmt.Errorf("could not start watcher in directory %s: %s", path, err)
-		}
+	om.listenerMap = map[string]*ServiceObsListener{}
+
+	for _, watcherStopCh := range om.watcherStopChMap {
+		watcherStopCh <- struct{}{}
 	}
-	// if not a directory and not a JSON, ignore
-	if filepath.Ext(stat.Name()) != ".json" {
-		return nil
+	om.watcherStopChMap = map[string]chan struct{}{}
+
+	om.metrics.observationsGauge.Set(0)
+	om.running = false
+}
+
+// ConfigMap returns a map of id:*ServiceObsConfig for all running observations.
+func (om *ServiceObsManager) ConfigMap() map[string]*ServiceObsConfig {
+	om.Lock()
+	defer om.Unlock()
+
+	obsMap := make(map[string]*ServiceObsConfig, len(om.listenerMap))
+	for id, obs := range om.listenerMap {
+		// copy value so that it can't be changed
+		obsMap[id] = &ServiceObsConfig{}
+		*obsMap[id] = *obs.config
 	}
 
-	// create new observation from json
-	obs, err := NewServiceObservationFromFile(path, s.opts, s.observationMetrics, s.reconnectCtr)
+	return obsMap
+}
+
+// Set creates or updates an observation
+// if an observation exists with the same ID, it is updated
+// otherwise, a new observation is created
+func (om *ServiceObsManager) Set(config *ServiceObsConfig) error {
```

Comment:
I don't think having user pass the ID is a good thing, it is generally a bit of an anti pattern. I would rather generate UID and return it in response.

Reply:
Add/Update aren't separate anymore though.  `Set` is an Upsert, where we must know which `ID` to check against

For the FS Watcher, it is using the File Path as the `ID`, so any File Path will only ever have 1 observation

I think that `Set` is working pretty well here for the FS Watcher, and for external integrations that have deterministic IDs for their observations.  If an external integration wants to do an Add/Update, they could make a helper function for `Add` that assigns and returns a random ID.  Then use `Set` for updates only.